crates/wasmtime/src/runtime/vm/gc/enabled/structref.rs

```diff
@@ -0,0 +1,471 @@
+            Val::I32(i) if ty.is_i16() => data.write_pod(offset, i as i16),
+            Val::I32(i) => data.write_pod(offset, i),
+            Val::I64(i) => data.write_pod(offset, i),
+            Val::F32(f) => data.write_pod(offset, f),
+            Val::F64(f) => data.write_pod(offset, f),
+            Val::V128(v) => data.write_pod(offset, v),
+
+            // For GC-managed references, we need to take care to run the
+            // appropriate barriers, even when we are writing null references
+            // into the struct.
+            //
+            // POD-read the old value into a local copy, run the GC write
+            // barrier on that local copy, and then POD-write the updated
+            // value back into the struct. This avoids transmuting the inner
+            // data, which would probably be fine, but this approach is
+            // Obviously Correct and should get us by for now. If LLVM isn't
+            // able to elide some of these unnecessary copies, and this
+            // method is ever hot enough, we can always come back and clean
+            // it up in the future.
+            Val::ExternRef(e) => {
+                let raw = data.read_pod::<u32>(offset);
+                let mut gc_ref = VMGcRef::from_raw_u32(raw);
+                let e = match e {
+                    Some(e) => Some(e.try_gc_ref(store)?.unchecked_copy()),
+                    None => None,
+                };
+                store.gc_store_mut()?.write_gc_ref(&mut gc_ref, e.as_ref());
+                let mut data = store.gc_store_mut()?.struct_data(self, layout.size);
+                data.write_pod(offset, gc_ref.map_or(0, |r| r.as_raw_u32()));
+            }
+            Val::AnyRef(a) => {
+                let raw = data.read_pod::<u32>(offset);
+                let mut gc_ref = VMGcRef::from_raw_u32(raw);
+                let a = match a {
+                    Some(a) => Some(a.try_gc_ref(store)?.unchecked_copy()),
+                    None => None,
+                };
+                store.gc_store_mut()?.write_gc_ref(&mut gc_ref, a.as_ref());
+                let mut data = store.gc_store_mut()?.struct_data(self, layout.size);
+                data.write_pod(offset, gc_ref.map_or(0, |r| r.as_raw_u32()));
+            }
+
+            Val::FuncRef(_) => todo!("funcrefs inside gc objects not yet implemented"),
+        }
+        Ok(())
+    }
+
+    /// Initialize a field in this structref that is currently uninitialized.
+    ///
+    /// Calling this method on a structref that has already had the associated
+    /// field initialized will result in GC bugs. These are memory safe but will
+    /// lead to generally incorrect behavior such as panics, leaks, and
+    /// incorrect results.
+    ///
+    /// Returns an error if `val` is a GC reference that has since been
+    /// unrooted.
+    ///
+    /// Panics on out-of-bounds accesses.
+    pub fn initialize_field(
```

Comment:
It took me a bit to work out why this exists given `write_field` above. I think I understand now in that it has to do with gc barriers and things like that. This leads me to a few questions:

* Could the documentation be updated to explicitly discuss the difference between this method and `write_field`?
* Would it be possible to change these to a single method? Perhaps with a boolean flag indicating if the previous data is "valid"?

Reply:
I can certainly add more docs clarifying their relationship.

I don't think I want to merge them into a single method, since if the method isn't inlined then that boolean argument will be materialized and branched on at runtime, which seems pretty silly since all call sites will be statically choosing one or the other. I guess I could have a `const bool` generic parameter... not sure that is actually any nicer than two different methods tho.

Reply:
None of this is perf-sensitive though, right? If perf matters there's tons of places we'd want to improve I think, chiefly being removing the usage of `Val`. Regardless though I think more docs is sufficient for now.

Reply:
It isn't perf-sensitive yet, but it will likely eventually be, yes.

I'd certainly like to add a `TypedStructRefPre` or something to that effect, but I didn't want to do too much too eagerly, especially before I even have the spec tests passing.